src/pipeline.cc

```diff
@@ -483,52 +520,53 @@ void LaunchStdin() {
   ThreadEnter();
   std::thread([]() {
     set_thread_name("stdin");
+    bool eof = false;
     std::string str;
-    const std::string_view kContentLength("Content-Length: ");
-    while (true) {
-      int len = 0;
+    while (!eof) {
+      HeadersData headers{};
+      ReadHeaders(&eof, &headers);
+      if (eof)
+        return;
+      if (headers.content_length == 0)
+        continue;
+
       str.clear();
-      while (true) {
+      str.reserve(headers.content_length);
+      for (std::size_t i = 0; i < headers.content_length; ++i) {
         int c = getchar();
-        if (c == EOF)
+        if (c == EOF) {
+          eof = true;
           return;
-        if (c == '\n') {
-          if (str.empty())
-            break;
-          if (!str.compare(0, kContentLength.size(), kContentLength))
-            len = atoi(str.c_str() + kContentLength.size());
-          str.clear();
-        } else if (c != '\r') {
-          str += c;
         }
+        str.push_back(c);
       }
 
-      str.resize(len);
-      for (int i = 0; i < len; ++i) {
-        int c = getchar();
-        if (c == EOF)
-          return;
-        str[i] = c;
-      }
-
-      auto message = std::make_unique<char[]>(len);
+      auto message = std::make_unique<char[]>(headers.content_length);
       std::copy(str.begin(), str.end(), message.get());
       auto document = std::make_unique<rapidjson::Document>();
-      document->Parse(message.get(), len);
-      assert(!document->HasParseError());
+      document->Parse(message.get(), headers.content_length);
+      if (document->HasParseError()) {
+        LOG_S(WARNING) << "Json parse error";
+        continue;
+      }
 
       JsonReader reader{document.get()};
       if (!reader.m->HasMember("jsonrpc") ||
-          std::string((*reader.m)["jsonrpc"].GetString()) != "2.0")
-        break;
+          std::string((*reader.m)["jsonrpc"].GetString()) != "2.0") {
+        LOG_S(WARNING) << "Bad jsonrpc member";
```

Comment:
The error checking for jsonrpc may not be that useful.

Reply:
I preserve previous behaviour AND log reason if the request is skipped